From a7388aa31cce6df7bdd544177c520a4d70fe1504 Mon Sep 17 00:00:00 2001
From: Ronghua Zeng <quic_ronghuaz@quicinc.com>
Date: Wed, 20 Dec 2023 15:04:10 +0800
Subject: [PATCH] ath9k: fix the tx queues blocking issue

In MCC mode, only the queues corresponding to the
current chanctx should be woken up after switching
to the next channel. To prevent blocking caused by
scheduling the incorrect queues to the active channel.

Signed-off-by: Ronghua Zeng <quic_ronghuaz@quicinc.com>
---
 drivers/net/wireless/ath/ath9k/main.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath9k/main.c b/drivers/net/wireless/ath/ath9k/main.c
index 9eefe82..eee7a54 100644
--- a/drivers/net/wireless/ath/ath9k/main.c
+++ b/drivers/net/wireless/ath/ath9k/main.c
@@ -276,7 +276,23 @@ static bool ath_complete_reset(struct ath_softc *sc, bool start)
 
 	ath9k_hw_set_interrupts(ah);
 	ath9k_hw_enable_interrupts(ah);
-	ieee80211_wake_queues(sc->hw);
+	if (ath9k_is_chanctx_enabled() && start) {
+		int i;
+
+		if (sc->cur_chan == &sc->offchannel.chan) {
+			ieee80211_wake_queue(sc->hw,
+					sc->hw->offchannel_tx_hw_queue);
+		} else {
+			for (i = 0; i < IEEE80211_NUM_ACS; i++)
+				ieee80211_wake_queue(sc->hw,
+						sc->cur_chan->hw_queue_base + i);
+		}
+
+		if (ah->opmode == NL80211_IFTYPE_AP)
+			ieee80211_wake_queue(sc->hw, sc->hw->queues - 2);
+	} else {
+		ieee80211_wake_queues(sc->hw);
+	}
 	ath9k_p2p_ps_timer(sc);
 
 	return true;
-- 
2.7.4

